### PR TITLE
Shared memory bitmap allocator library

### PR DESCRIPTION
**Pull request by evanstella**
### Summary of this Pull Request (PR)

Implementation of a shared memory bitmap allocator library and memmgr aligned allocations.

### Intent for your PR

Choose one (Mandatory):

- [x] This PR is for a code-review and is intended to get feedback, but not to be pulled yet.
- [ ] This PR is mature, and ready to be integrated into the repo.

### Reviewers (Mandatory):
@gparmer 

### Code Quality

As part of this pull request, I've considered the following:

[Style](https://github.com/gparmer/composite/raw/ppos/doc/style_guide/composite_coding_style.pdf):

- [x] Comments adhere to the Style Guide (SG)
- [x] Spacing adhere's to the SG
- [x] Naming adhere's to the SG
- [x] All other aspects of the SG are adhered to, or exceptions are justified in this pull request
- [ ] I have run the auto formatter on my code before submitting this PR (see doc/auto_formatter.md for instructions)

[Code Craftsmanship](http://www2.seas.gwu.edu/~gparmer/posts/2016-03-07-code-craftsmanship.html):

- [x] I've made an attempt to remove all redundant code
- [x] I've considered ways in which my changes might impact existing code, and cleaned it up
- [x] I've formatted the code in an effort to make it easier to read (proper error handling, function use, etc...)
- [x] I've commented appropriately where code is tricky
- [x] I agree that there is no "throw-away" code, and that code in this PR is of high quality

### Testing

I've tested the code using the following test programs (provide list here):

- [ ] micro_booter
- [x] unit_pingpong
- [x] unit_schedtests
- [x] unit_pingshmem (included)

### CODE REVIEW

File: src/components/lib/shm_bm/shm_bm.c [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+	// obj has not been allocated
+	if ((SHM_BM_REFC(shm) + id) == 0) return 0;
+
+	cos_faab(SHM_BM_REFC(shm) + id, 1);
+	return SHM_BM_DATA(shm) + (id * SHM_BM_SIZE(shm));
+}
+
+void *   
+shm_bm_obj_take(shm_bm_t shm, shm_bufid_t id)
```
Comment (gparmer) [inline src/components/lib/shm_bm/shm_bm.c:158]: Kinda confused as I figured there would be some modification to the structure to refcnt. I suppose this might be the optimization that for passing ownership.

File: src/components/lib/shm_bm/shm_bm.c [hunk 2]
```diff
@@ -0,0 +1,193 @@
+#include <shm_bm.h>
+#include <limits.h>
+#include <consts.h>
+#include <string.h>
+
+#define SHM_BM_BITMAP_BLOCK (sizeof (word_t) * CHAR_BIT)
+#define SHM_BM_BITS_TO_WORDS(bits) (round_up_to_pow2(bits, SHM_BM_BITMAP_BLOCK) / SHM_BM_BITMAP_BLOCK)
+#define SHM_BM_ALLOC_ALIGNMENT 1ul << 22 // sm regions are on 4MB boundries
```
Comment (gparmer) [inline src/components/lib/shm_bm/shm_bm.c:8]: You can't know that `<<` has a higher associativity than all places this macro is used. Add `()`

File: src/components/lib/shm_bm/shm_bm.c [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+void
+shm_bm_obj_free(void *ptr)
+{
+	int      bit, index, offset;
+	shm_bm_t shm;
+	word_t  *bm;
+
+	// mask out the bits less significant than the allocation alignment
+	shm = (shm_bm_t) ((word_t) ptr & ~((SHM_BM_ALLOC_ALIGNMENT >> 1) - 1));
```
Comment (gparmer) [inline src/components/lib/shm_bm/shm_bm.c:176]: I don't understand the `>> 1` here. I'd think this is correct without that.

File: src/components/lib/shm_bm/shm_bm.c [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+	bit = ((byte_t *) ptr - SHM_BM_DATA(shm)) / SHM_BM_SIZE(shm);
```
Comment (gparmer) [inline src/components/lib/shm_bm/shm_bm.c:178]: If we stored the size of allocations in a pow-of-two order, then this could be a rightshift rather than a general division. I'm think it is OK to waste the memory, and track object sizes as power-of-two-order instead.

File: src/components/lib/shm_bm/shm_bm.c [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+	if (bit < 0 || bit >= SHM_BM_NOBJ(shm)) return;
+
+	if (cos_faab(SHM_BM_REFC(shm) + bit, -1) > 1)
+		return;
+
+	// droping the last reference, must free obj
+	index  = bit / SHM_BM_NOBJ(shm);
```
Comment (gparmer) [inline src/components/lib/shm_bm/shm_bm.c:185]: My brain is not letting me understand the `SHM_BM_NOBJ` logic here and in the next lines.

Feels like this should all be dividing by word size at some point to find the word in the bitmap, but I don't see that.

File: src/components/lib/shm_bm/shm_bm.c [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+ * (bitmap, ref counts, data) because we cannot store pointer to 
+ * them in the header since this memory is shared across virt
+ * address spaces...
+ */
+#define SHM_BM_BITM(shm) ((word_t *)  (shm + ((struct shm_bm_header *) shm)->bitm_offset))
+#define SHM_BM_REFC(shm) ((refcnt_t *)(shm + ((struct shm_bm_header *) shm)->refc_offset))
+#define SHM_BM_DATA(shm) ((byte_t *)  (shm + ((struct shm_bm_header *) shm)->data_offset))
+
+// ...and to make the code a little cleaner when using the opaque shm_bm_t type
+#define SHM_BM_SIZE(shm) (((struct shm_bm_header *) shm)->objsz)
```
Comment (gparmer) [inline src/components/lib/shm_bm/shm_bm.c:40]: This is hard to follow. Maybe `SHM_BM_OBJSZ`?

File: src/components/lib/shm_bm/shm_bm.c (hunk 5 above)
Comment (gparmer) [reply src/components/lib/shm_bm/shm_bm.c:185]: I don't really see how the comment relates to the code here.

File: src/components/lib/shm_bm/shm_bm.c (hunk 4 above)
Comment (gparmer) [reply src/components/lib/shm_bm/shm_bm.c:178]: `bit` doesn't strike me as descriptive here. `obj_idx` maybe?

File: src/components/lib/shm_bm/shm_bm.c [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+	shm_bm_t shm;
+	word_t  *bm;
+
+	// mask out the bits less significant than the allocation alignment
+	shm = (shm_bm_t) ((word_t) ptr & ~((SHM_BM_ALLOC_ALIGNMENT >> 1) - 1));
+
+	bit = ((byte_t *) ptr - SHM_BM_DATA(shm)) / SHM_BM_SIZE(shm);
+	if (bit < 0 || bit >= SHM_BM_NOBJ(shm)) return;
+
+	if (cos_faab(SHM_BM_REFC(shm) + bit, -1) > 1)
```
Comment (gparmer) [inline src/components/lib/shm_bm/shm_bm.c:181]: The pointer arith is necessary through a lot of this, but not for looking up the reference counts.

File: src/kernel/include/shared/util.h [hunk 8]
```diff
@@ -30,6 +30,16 @@ cos_faa(int *var, int value)
 	return value;
 }
 
+static inline unsigned char
```
Comment (gparmer) [inline src/kernel/include/shared/util.h:33]: document this? I don't know what it does ;-)

**Review by gparmer — COMMENTED** [review]
Really amazing code, thanks! Some tweaks and questions. Can chat in meeting.

File: src/components/lib/shm_bm/shm_bm.c (hunk 1 above)
Comment (evanstella) [reply src/components/lib/shm_bm/shm_bm.c:158]: Yes, this is documented in the header file; im trying to think of a better name than "take"

File: src/components/lib/shm_bm/shm_bm.c (hunk 7 above)
Comment (evanstella) [reply src/components/lib/shm_bm/shm_bm.c:181]: I think it is? `SHM_BM_REFC(shm) + bit` gets `&refcnt[bit]`. FAA takes a pointer so we need to do the arithmetic to get to the pointer of the refcnt we are trying to decrement
